include/host_utils.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "host_queue.h"
+#include "item.h"
+
+namespace celerity {
+
+template <int Dims, typename Functor>
+void for_each_item(const partition<Dims> part, Functor func) {
+	static_assert(std::is_invocable_v<Functor, item<Dims>>, "for_each_item functor needs to be invocable with celerity::item<Dims>.");
+
+	if constexpr(Dims == 0) { //
+		func(detail::make_item<0>(id<0>(), id<0>(), range<0>()));
+	} else if constexpr(Dims == 1) {
+		for(size_t d0 = 0; d0 < part.get_subrange().range[0]; ++d0) {
+			func(detail::make_item<1>(id<1>(d0) + part.get_subrange().offset, part.get_subrange().offset, part.get_global_size()));
+		}
+	} else if constexpr(Dims == 2) {
+		for(size_t d0 = 0; d0 < part.get_subrange().range[0]; ++d0) {
+			for(size_t d1 = 0; d1 < part.get_subrange().range[1]; ++d1) {
+				func(detail::make_item<2>(id<2>(d0, d1) + part.get_subrange().offset, part.get_subrange().offset, part.get_global_size()));
+			}
+		}
+	} else if constexpr(Dims == 3) {
+		for(size_t d0 = 0; d0 < part.get_subrange().range[0]; ++d0) {
+			for(size_t d1 = 0; d1 < part.get_subrange().range[1]; ++d1) {
+				for(size_t d2 = 0; d2 < part.get_subrange().range[2]; ++d2) {
+					func(detail::make_item<3>(id<3>(d0, d1, d2) + part.get_subrange().offset, part.get_subrange().offset, part.get_global_size()));
+				}
+			}
+		}
+	}
+}
+
+template <int Dims, typename Functor>
+void for_each_item(const range<Dims> range, Functor func) {
```

Comment:
This allows the user to summon the "magic" type `celerity::item` for the first time. Could we alternatively iterate over `id`s instead of items, or if items are necessary for this to work, stage the change through `namespace celerity::experimental`?

Reply:
I'm not sure I understand the problem here: it iterates over `item`s because that's what `parallel_for` does and the idea is for this to have the equivalent signature.

Reply:
Fom the user's perspective, so far the only way to create a `celerity::item` was to launch a kernel. `nd_item` is a "magic type" because of its synchronization functionality, and although `item` is effectively just a struct, I would categorize it to be "equally magic" and something only the runtime should be able to create.

In contrast, the `range` overload of `for_each_item` allows users to create `item`s wherever they want. This is not a problem for the `partition` overload, because a `partition` is not user-constructible.

Reply:
I see what you mean now, but I don't feel like having this new signature makes a practical difference: the user still isn't creating the item, it's just handed to the functor they provide by a Celerity function.  
And I don't think we are introducing any new significant constraints on what a `celerity::item` can be, since its already part of an external interface (arguable of *the* most important external interface in Celerity).

The reason I'd like to keep this interface the same is that this way, it is a 1:1 match with `parallel_for`, which I feel is more natural than iterating over something else here.

Reply:
> [or] stage the change through namespace celerity::experimental

I think this would be a good idea regardless, until we have a real-world use case for it.

Reply:
I've now put the new interface into `celerity::experimental`.